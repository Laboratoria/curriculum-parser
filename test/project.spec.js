const path = require('path');
const mongoose = require('mongoose');
const models = require('@laboratoria/models')(mongoose);
const helpers = require('./helpers');
const project = require('../lib/project');
const pkg = require('../package.json');


describe('project', () => {
  it('should reject when project dir not in expected format', () => {
    const p = helpers.resolveFixturePath('a-project');
    expect.assertions(2);
    return project(p, models)
      .catch((err) => {
        expect(err.message).toBe('Expected project dir to be in 00-slug format and got a-project');
        expect(err.path).toBe(p);
      });
  });

  it('should reject when language not supported', () => {
    const p = helpers.resolveFixturePath('01-foo');
    expect.assertions(1);
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'en-GB',
    })
      .catch((err) => {
        expect(err.message).toBe('Unsupported language: en');
      });
  });

  it('should reject when dir doesnt exist', () => {
    expect.assertions(2);
    return project('01-foo', models, { locale: 'es-ES' })
      .catch((err) => {
        expect(err.message).toMatch(/no such file or directory/);
        expect(err.code).toBe('ENOENT');
      });
  });

  it('should reject when README.md is empty', () => {
    const p = helpers.resolveFixturePath('00-course-empty');
    expect.assertions(2);
    return project(p, models, { locale: 'es-ES' })
      .catch((err) => {
        expect(err.message).toBe('Project README.md is empty');
        expect(err.path).toBe(path.join(p, 'README.md'));
      });
  });

  it('should reject when README.md doesnt start with h1', () => {
    const p = helpers.resolveFixturePath('01-a-project-without-a-title');
    const p2 = helpers.resolveFixturePath('01-a-project-without-a-bad-title');
    return project(p, models, { locale: 'es-ES' })
      .then(() => {
        throw new Error('This should never happen');
      })
      .catch((err) => {
        expect(err.message).toBe('Expected README.md to start with h1 and instead saw heading (depth: 2)');
        expect(err.path).toBe(path.join(p, 'README.md'));

        return project(p2, models, { locale: 'es-ES' });
      })
      .then(() => {
        throw new Error('This should never happen');
      })
      .catch((err) => {
        expect(err.message).toBe('Expected README.md to start with h1 and instead saw paragraph');
        expect(err.path).toBe(path.join(p2, 'README.md'));
      });
  });

  it('should parse portuguese project', () => {
    const p = helpers.resolveFixturePath('01-a-project-with-pt-translation');
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'pt-BR',
      suffix: 'pt',
    })
      .then(({ createdAt, ...parsed }) => {
        expect(parsed.slug).toBe('a-project-with-pt-translation-pt');
        expect(parsed.locale).toBe('pt-BR');
        expect(parsed).toMatchSnapshot();
      });
  });

  it('should reject when unknown learning objective', () => {
    const p = helpers.resolveFixturePath('01-a-project-with-unknown-learning-objective');
    expect.assertions(2);
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'es-ES',
      lo: path.join(__dirname, 'fixtures', 'learning-objectives'),
    })
      .catch((err) => {
        expect(err.message).toBe('Unknown learning objectives: html/foo.');
        expect(err.path).toBe(path.join(p, 'README.md'));
      });
  });

  it('should parse a project with learning objectives without validating against known list', () => {
    const p = helpers.resolveFixturePath('01-a-project-with-learning-objectives');
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'es-ES',
    })
      .then(({ parserVersion, createdAt, ...parsed }) => {
        expect(parserVersion).toBe(pkg.version);
        expect(parsed).toMatchSnapshot();
      });
  });

  it('should parse a project with learning objectives validating against known list', () => {
    const p = helpers.resolveFixturePath('01-a-project-with-learning-objectives');
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'es-ES',
      lo: path.join(__dirname, 'fixtures', 'learning-objectives'),
    })
      .then(({ parserVersion, createdAt, ...parsed }) => {
        expect(parserVersion).toBe(pkg.version);
        expect(parsed).toMatchSnapshot();
      });
  });

  it('should parse a project with learning objectives', () => {
    const p = helpers.resolveFixturePath('01-a-project-with-learning-objectives');
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'es-ES',
    })
      .then(({ parserVersion, createdAt, ...parsed }) => {
        expect(parserVersion).toBe(pkg.version);
        expect(parsed).toMatchSnapshot();
      });
  });

  it('should expand learning objectives children when only parent is mentioned', () => {
    const p = helpers.resolveFixturePath('01-a-project-with-lo-needing-expansion');
    return project(p, models, {
      track: 'js',
      repo: 'Laboratoria/bootcamp',
      version: '1.0.0',
      locale: 'es-ES',
      lo: path.join(__dirname, 'fixtures', 'learning-objectives'),
    })
      .then((result) => {
        expect(result.learningObjectives).toMatchSnapshot();
      });
  });
});
