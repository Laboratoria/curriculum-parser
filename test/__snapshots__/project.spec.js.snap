// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project should expand learning objectives children when only parent is mentioned 1`] = `
Array [
  "html/semantics",
  "css/selectors",
  "css/flexbox",
  "css/grid",
  "css/media-queries",
  "js/async/callbacks",
  "js/async/promises",
  "js/data-types/strings",
  "scm/git/setup",
  "scm/git/intro",
  "scm/git/integration",
  "scm/github/setup",
  "scm/github/gh-pages",
  "scm/github/collaboration",
  "scm/github/project-management",
  "scm/github/ci",
  "http/headers",
]
`;

exports[`project should parse a project with learning objectives 1`] = `
Object {
  "learningObjectives": Array [
    "html/semantics",
    "css/selectors",
    "dom/selectors",
    "dom/events",
    "dom/manipulation",
    "js/data-types/primitive-vs-non-primitive",
    "js/data-types/strings",
    "js/variables",
    "js/conditionals",
    "js/functions",
    "js/semantics",
    "ux/user-understanding",
    "ux/prototyping",
  ],
  "locale": "es-ES",
  "path": "test/fixtures/01-a-project-with-learning-objectives",
  "prefix": 1,
  "repo": "Laboratoria/bootcamp",
  "slug": "a-project-with-learning-objectives",
  "title": "A project",
  "track": "js",
  "version": "1.0.0",
}
`;

exports[`project should parse a project with learning objectives validating against known list 1`] = `
Object {
  "learningObjectives": Array [
    "html/semantics",
    "css/selectors",
    "dom/selectors",
    "dom/events",
    "dom/manipulation",
    "js/data-types/primitive-vs-non-primitive",
    "js/data-types/strings",
    "js/variables/declaration",
    "js/conditionals",
    "js/functions",
    "js/semantics",
    "ux/user-understanding",
    "ux/prototyping",
  ],
  "locale": "es-ES",
  "path": "test/fixtures/01-a-project-with-learning-objectives",
  "prefix": 1,
  "repo": "Laboratoria/bootcamp",
  "slug": "a-project-with-learning-objectives",
  "title": "A project",
  "track": "js",
  "version": "1.0.0",
}
`;

exports[`project should parse a project with learning objectives without validating against known list 1`] = `
Object {
  "learningObjectives": Array [
    "html/semantics",
    "css/selectors",
    "dom/selectors",
    "dom/events",
    "dom/manipulation",
    "js/data-types/primitive-vs-non-primitive",
    "js/data-types/strings",
    "js/variables",
    "js/conditionals",
    "js/functions",
    "js/semantics",
    "ux/user-understanding",
    "ux/prototyping",
  ],
  "locale": "es-ES",
  "path": "test/fixtures/01-a-project-with-learning-objectives",
  "prefix": 1,
  "repo": "Laboratoria/bootcamp",
  "slug": "a-project-with-learning-objectives",
  "title": "A project",
  "track": "js",
  "version": "1.0.0",
}
`;

exports[`project should parse portuguese project 1`] = `
Object {
  "learningObjectives": undefined,
  "locale": "pt-BR",
  "parserVersion": "3.0.0-alpha.1",
  "path": "test/fixtures/01-a-project-with-pt-translation",
  "prefix": 1,
  "repo": "Laboratoria/bootcamp",
  "slug": "a-project-with-pt-translation-pt",
  "title": "Cifrado CÃ©sar",
  "track": "js",
  "version": "1.0.0",
}
`;
